programs/smart-wallet/src/instructions/buffer_execute_ix.rs

```diff
@@ -0,0 +1,52 @@
+//! Executes an instruction off of the [InstructionBuffer].
+
+use crate::*;
+
+use anchor_lang::solana_program::program::invoke_signed;
+
+#[derive(Accounts)]
+pub struct ExecuteBufferIX<'info> {
+    #[account(mut)]
+    pub buffer: Account<'info, InstructionBuffer>,
+    /// The buffer's [SmartWallet]
+    pub smart_wallet: Account<'info, SmartWallet>,
+    /// The [InstructionBuffer::executer].
+    pub executor: Signer<'info>,
+}
+
+pub fn handler<'info>(ctx: Context<'_, '_, '_, 'info, ExecuteBufferIX<'info>>) -> Result<()> {
+    let smart_wallet = &ctx.accounts.smart_wallet;
+    let wallet_seeds: &[&[&[u8]]] = &[&[
+        b"GokiSmartWallet" as &[u8],
+        &smart_wallet.base.to_bytes(),
+        &[smart_wallet.bump],
+    ]];
+
+    let buffer = &mut ctx.accounts.buffer;
+    let ix = &buffer.instructions[buffer.exec_count as usize];
+    invoke_signed(&ix.into(), ctx.remaining_accounts, wallet_seeds)?;
+
+    buffer.exec_count += 1;
+
+    Ok(())
+}
+
+impl<'info> Validate<'info> for ExecuteBufferIX<'info> {
+    fn validate(&self) -> Result<()> {
+        assert_keys_eq!(self.buffer.executer, self.executor.key());
+        assert_keys_eq!(self.buffer.smart_wallet, self.smart_wallet.key());
+
+        invariant!(self.buffer.finalized_at > 0, BufferNotFinalized);
+        invariant!(usize::from(self.buffer.exec_count) < self.buffer.instructions.len());
```

Comment:
Move this into a method on `InstructionBuffer`, called `has_fully_executed`.

Also, this should throw an `ErrorCode` for better UX.